staging/src/k8s.io/client-go/tools/record/v1beta1/eventseries.go

```diff
@@ -0,0 +1,411 @@
+	NewRecorder(scheme *runtime.Scheme, reportingController string) EventRecorder
+}
+
+type recorderImpl struct {
+	scheme              *runtime.Scheme
+	reportingController string
+	reportingInstance   string
+	*watch.Broadcaster
+	clock clock.Clock
+}
+
+type eventBroadcasterImpl struct {
+	*watch.Broadcaster
+	onGoingSeries   []*v1beta1.Event
+	onGoingSeriesMu sync.RWMutex
+	sleepDuration   time.Duration
+	sink            EventSink
+}
+
+// EventSink knows how to store events (client.Client implements it.)
+// EventSink must respect the namespace that will be embedded in 'event'.
+// It is assumed that EventSink will return the same sorts of errors as
+// pkg/client's REST client.
+type EventSink interface {
+	Create(event *v1beta1.Event) (*v1beta1.Event, error)
+	Update(event *v1beta1.Event) (*v1beta1.Event, error)
+	Patch(oldEvent *v1beta1.Event, data []byte) (*v1beta1.Event, error)
+	List(opts metav1.ListOptions) (*v1beta1.EventList, error)
+}
+
+// NewBroadcaster Creates a new event broadcaster.
+func NewBroadcaster(sink EventSink) EventBroadcaster {
+	return &eventBroadcasterImpl{
+		Broadcaster:   watch.NewBroadcaster(maxQueuedEvents, watch.DropIfChannelFull),
+		onGoingSeries: []*v1beta1.Event{},
+		sleepDuration: defaultSleepDuration,
+		sink:          sink,
+	}
+}
+
+// NewBroadcasterForTestSeries Creates a new event broadcaster for test purposes.
+func NewBroadcasterForTestSeries(sink EventSink, sleepDuration time.Duration, onGoingSeries []*v1beta1.Event) EventBroadcaster {
+	return &eventBroadcasterImpl{
+		Broadcaster:   watch.NewBroadcaster(maxQueuedEvents, watch.DropIfChannelFull),
+		onGoingSeries: onGoingSeries,
+		sleepDuration: sleepDuration,
+		sink:          sink,
+	}
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) eventThreshold() {
+	eventBroadcaster.onGoingSeriesMu.RLock()
+	defer eventBroadcaster.onGoingSeriesMu.RUnlock()
+	for _, event := range eventBroadcaster.onGoingSeries {
+
+		if event.Series != nil {
+			recordEvent(eventBroadcaster.sink, event, true)
+		}
+
+	}
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) checkEndingSeries() {
+	eventBroadcaster.onGoingSeriesMu.Lock()
+	defer eventBroadcaster.onGoingSeriesMu.Unlock()
+	for i := len(eventBroadcaster.onGoingSeries) - 1; i >= 0; i-- {
+		eventSerie := eventBroadcaster.onGoingSeries[i].Series
+		if eventSerie != nil {
+
+			if eventSerie.LastObservedTime.Time.Add(6 * time.Minute).Before(time.Now()) {
+
+				eventSerie.State = v1beta1.EventSeriesStateFinished
+
+				if recordEvent(eventBroadcaster.sink, eventBroadcaster.onGoingSeries[i], true) {
+					eventBroadcaster.onGoingSeries = append(eventBroadcaster.onGoingSeries[:i], eventBroadcaster.onGoingSeries[i+1:]...)
+				}
+			}
+		}
+
+	}
+}
+
+func isIsomorphic(onGoingSeries []*v1beta1.Event, newEvent *v1beta1.Event) (bool, int) {
+
+	for position, eventSerie := range onGoingSeries {
+
+		if eventSerie.Action == newEvent.Action && eventSerie.Reason == newEvent.Reason &&
```

Comment:
@wojtek-t  - We might want to optimize this by using a map with a key that is hash of the elements that makes an event isomorphic. performance-wise, computing the hash + o(1) for map access should cost less than o(n) with a slice